### ISSUE TITLE
Rule for Unnecessary Optional Chaining 

### ISSUE THREAD

**Issue by nicholascm** [eslint/eslint#12139]
**Please describe what the rule should do:**

The rule will alert to the presence of unnecessary optional chaining / safe navigation calls.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[x] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js

// EXAMPLE 1
let  test = {};

// since variable `test` is guaranteed to  be defined as an object, no need for safe navigation off of `test`

let prop = test?.prop?.somethingElse; 

// EXAMPLE 2
let example = { 
   nested: { 
       prop: { }
    }
}

// provides warning that there is no need for optional chaining
const deepest = example?.nested?.prop;

// EXAMPLE 3
const copy = item ? cloneDeep(item) : {};

// provides warning that no need for optional chaining after `copy` as it is guaranteed to be defined
copy.list = copy?.list?.map(p => SomeService.create(p)) || [];

```

**Why should this rule be included in ESLint (instead of a plugin)?**

As optional chaining has become stage 3 this will be more and more common. Adding unecessary optional chaining operators causes the production of unnecessary statements in the transpiled code. 

![image](https://user-images.githubusercontent.com/8185671/63446053-82d01180-c407-11e9-85dc-e3b9f20c54e8.png)

**Are you willing to submit a pull request to implement this rule?**

Looking into that...would appreciate any help also.

**Comment by platinumazure** [comment]
Hi @nicholascm, thanks for the issue.

Our policy for core ESLint rules is to only follow the latest standard (Stage 4) language features. (More info in our [README](https://github.com/eslint/eslint/blob/master/README.md#what-about-experimental-features).) So I don't think this is something we can accept at this time.

I would suggest making a feature request in `eslint-plugin-babel` if you think this might be a good idea for that ecosystem.

**Comment by ljharb** [comment]
Either way, this would only be useful when you'd used optional chaining on an object literal defined in the same file, and there's zero harm in using it in that case (because engines could optimize it out) - I'm not sure this is widely applicable.

**Comment by platinumazure** [comment]
Good point.

If Typescript supports optional chaining, it could be good to open an issue in [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint).
